test/e2e/nodepool_machineconfig_test.go

```diff
@@ -0,0 +1,204 @@
+	e2eutil.WaitForNReadyNodes(t, testContext, guestClient, numNodes)
+
+	// Wait for the rollout to be complete
+	t.Logf("Waiting for cluster rollout. Image: %s", globalOpts.LatestReleaseImage)
+	e2eutil.WaitForImageRollout(t, testContext, client, guestClient, hostedCluster, globalOpts.LatestReleaseImage)
+	err = client.Get(testContext, crclient.ObjectKeyFromObject(hostedCluster), hostedCluster)
+	g.Expect(err).NotTo(HaveOccurred(), "failed to get hostedcluster")
+
+	ignitionConfig := ignitionapi.Config{
+		Ignition: ignitionapi.Ignition{
+			Version: "3.2.0",
+		},
+		Storage: ignitionapi.Storage{
+			Files: []ignitionapi.File{{
+				Node:          ignitionapi.Node{Path: "/etc/custom-config"},
+				FileEmbedded1: ignitionapi.FileEmbedded1{Contents: ignitionapi.Resource{Source: utilpointer.String("data:,content%0A")}},
+			}},
+		},
+	}
+	serializedIgnitionConfig, err := json.Marshal(ignitionConfig)
+	if err != nil {
+		t.Fatalf("failed to serialize ignition config: %v", err)
+	}
+	machineConfig := &mcfgv1.MachineConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   "custom",
+			Labels: map[string]string{"machineconfiguration.openshift.io/role": "worker"},
+		},
+		Spec: mcfgv1.MachineConfigSpec{Config: runtime.RawExtension{Raw: serializedIgnitionConfig}},
+	}
+	gvk, err := apiutil.GVKForObject(machineConfig, hyperapi.Scheme)
+	if err != nil {
+		t.Fatalf("failed to get typeinfo for %T from scheme: %v", machineConfig, err)
+	}
+	machineConfig.SetGroupVersionKind(gvk)
+	serializedMachineConfig, err := yaml.Marshal(machineConfig)
+	if err != nil {
+		t.Fatalf("failed to serialize machineConfig: %v", err)
+	}
+	machineConfigConfigMap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "custom-machine-config",
+			Namespace: hostedCluster.Namespace,
+		},
+		Data: map[string]string{"config": string(serializedMachineConfig)},
+	}
+	if err := client.Create(ctx, machineConfigConfigMap); err != nil {
+		t.Fatalf("failed to create configmap for custom machineconfig: %v", err)
+	}
+
+	nodepools := &hyperv1.NodePoolList{}
+	if err := client.List(ctx, nodepools, crclient.InNamespace(hostedCluster.Namespace)); err != nil {
+		t.Fatalf("failed to list nodepools in namespace %s: %v", hostedCluster.Namespace, err)
+	}
+
+	for _, nodepool := range nodepools.Items {
+		if !ownedBy(hostedCluster.UID, nodepool.OwnerReferences) {
+			continue
+		}
+		nodepool.Spec.Config = append(nodepool.Spec.Config, corev1.LocalObjectReference{Name: machineConfigConfigMap.Name})
+		if err := client.Update(ctx, &nodepool); err != nil {
+			t.Fatalf("failed to update nodepool %s after adding machineconfig: %v", nodepool.Name, err)
+		}
+	}
+
+	ds := machineConfigUpdatedVerificationDS.DeepCopy()
+	if err := guestClient.Create(ctx, ds); err != nil {
+		t.Fatalf("failed to create %s DaemonSet in guestcluster: %v", ds.Name, err)
+	}
+
+	t.Logf("waiting for rollout of updated nodepools")
+	err = wait.PollUntil(5*time.Second, func() (bool, error) {
```

Comment:
can we have a timeout here?

Also, to avoid having timeouts scattered/varying across tests might want to have a WaitFor* function in util containing the `PollImmediateWithContext` that this test can use. We are currently doing that for any other WaitFor*.

Reply:
Also as I was looking at this I realised WaitForNReadyNodes need some stronger guarantee as it's right now, the number of ready nodes could be = N accidentally in the middle of an ongoing roll out, so me might want to gate on a condition as well, I'll have a look separately.